rust/core-lib/src/edit_types.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2018 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A bunch of boilerplate for converting the `EditNotification`s we receive
+//! from the client into the events we use internally.
+//!
+//! This simplifies code elsewhere, and makes it easier to route events to
+//! the editor or view as appropriate.
+
+use movement::Movement;
+use ::rpc::{GestureType, LineRange, EditNotification, MouseAction};
+
+
+/// Events that only modify view state
+pub(crate) enum ViewEvent {
+    Move(Movement),
+    ModifySelection(Movement),
+    SelectAll,
+    Scroll(LineRange),
+    AddSelectionAbove,
+    AddSelectionBelow,
+    Click(MouseAction),
+    Drag(MouseAction),
+    Gesture { line: u64, col: u64, ty: GestureType },
+    GotoLine { line: u64 },
+    FindNext { wrap_around: Option<bool>, allow_same: Option<bool> },
+    FindPrevious { wrap_around: Option<bool> },
+    Cancel,
+}
+
+/// Events that modify the buffer
+pub(crate) enum BufferEvent {
+    Delete(Movement),
+    Backspace,
+    Transpose,
+    Undo,
+    Redo,
+    Uppercase,
+    Lowercase,
+    Indent,
+    Outdent,
+    Insert(String),
+    InsertNewline,
+    InsertTab,
+    Yank,
+}
+
+/// An event that needs special handling
+pub(crate) enum SpecialEvent {
+    DebugRewrap,
+    DebugPrintSpans,
+    RequestLines(LineRange),
+}
+
+pub(crate) enum EventDomain {
+    View(ViewEvent),
+    Buffer(BufferEvent),
+    Special(SpecialEvent),
+}
+
+impl From<BufferEvent> for EventDomain {
+    fn from(src: BufferEvent) -> EventDomain {
+        EventDomain::Buffer(src)
+    }
+}
+
+impl From<ViewEvent> for EventDomain {
+    fn from(src: ViewEvent) -> EventDomain {
+        EventDomain::View(src)
+    }
+}
+
+impl From<SpecialEvent> for EventDomain {
+    fn from(src: SpecialEvent) -> EventDomain {
+        EventDomain::Special(src)
+    }
+}
+
+impl From<EditNotification> for EventDomain {
+    fn from(src: EditNotification) -> EventDomain {
+        use self::EditNotification::*;
+        match src {
+            Insert { chars } =>
+                BufferEvent::Insert(chars).into(),
+            DeleteForward =>
+                BufferEvent::Delete(Movement::Left).into(),
```

Comment:
There is a bug here - DeleteForward should move _Right_ instead of _Left_.

Reply:
Nice, good spot. What do you mean about the tab character?

Reply:
DeleteBackward (backspace) detects and deleted a tab character if there is one. It used to be that DeleteForward also did the same, but it doesn't do so now.

Reply:
oooooh, interesting. You mean if there was four spaces from auto-indent, a single backspace would delete them all? I didn't think that forward delete ever did that, I thought it used different logic. it's possible I broke that though, will double check tomorrow.

Reply:
Yes, IIRC this is also how it works in other editors as well.